File: src/Avalonia.Samples/CompleteApps/Avalonia.MusicStore/ViewModels/MainWindowViewModel.cs
using System.Collections.ObjectModel;
using System.Linq;
using System.Threading.Tasks;
using Avalonia.MusicStore.Messages;
using Avalonia.MusicStore.Models;
using CommunityToolkit.Mvvm.ComponentModel;
using CommunityToolkit.Mvvm.Input;
using CommunityToolkit.Mvvm.Messaging;

namespace Avalonia.MusicStore.ViewModels
{
    public partial class MainWindowViewModel : ObservableObject
    {
        public ObservableCollection<AlbumViewModel> Albums { get; } = new();

        public MainWindowViewModel()
        {
            LoadAlbums();
        }

        /// <summary>
        /// This relay command send a message to initiate album purchase, adds the result to the collection and saves it to disk.
        /// </summary>
        [RelayCommand]
        private async Task AddAlbumAsync()
        {
            var album = await WeakReferenceMessenger.Default.Send(new PurchaseAlbumMessage());
            if (album is not null)
            {
                Albums.Add(album);
                await album.SaveToDiskAsync();
            }
        }

        /// <summary>
        /// Loads albums and their covers from cache.
        /// </summary>
        private async void LoadAlbums()
        {
            var albums = (await Album.LoadCachedAsync()).Select(x => new AlbumViewModel(x)).ToList();
            foreach (var album in albums)
            {
                Albums.Add(album);
            }
            var coverTasks = albums.Select(album => album.LoadCover());
            await Task.WhenAll(coverTasks);
        }
    }
}

File: src/Avalonia.Samples/FSharp/CompleteApps/MusicStore/ViewModels/MainWindowViewModel.fs
namespace MusicStore.ViewModels

open System.Collections.ObjectModel
open System.Diagnostics
open System.Threading.Tasks
open System.Windows.Input
open CommunityToolkit.Mvvm.ComponentModel
open CommunityToolkit.Mvvm.Input
open CommunityToolkit.Mvvm.Messaging
open MusicStore.Models

type MainWindowViewModel() =
    inherit ViewModelBase()

    [<ObservableProperty>]
    member val Albums = ObservableCollection<AlbumViewModel>() with get, set

    // Called by the View once it's opened
    member this.InitializeAsync() =
        this.LoadAlbumsAsync()

    /// Expose async commands to XAML (Buttons bind to ICommand)
    member this.AddAlbumCommand =
        AsyncRelayCommand(fun () ->
            task {
                Debug.WriteLine "AddAlbumCommand started"

                // Send the message to the previously registered handler and await the selected album
                let request = MusicStore.Messages.PurchaseAlbumMessage.Self()
                WeakReferenceMessenger.Default.Send(request) |> ignore

                let! albumOpt = request.Response

                match albumOpt with
                | None -> Debug.WriteLine "AddAlbumCommand got no album"
                | Some album ->
                    Debug.WriteLine $"AddAlbumCommand got album: %A{album.Title}"
                    this.Albums.Add album
                    do! album.SaveToDiskAsync()

                Debug.WriteLine "AddAlbumCommand completed"
            }
            :> Task)
        :> ICommand

    member private this.LoadAlbumsAsync() =
        task {
            let! albums = Album.LoadCachedAsync()
            Debug.WriteLine $"Loaded %d{Seq.length albums} cached albums"
            let albumViewModels = albums |> Seq.map AlbumViewModel |> Seq.toArray

            albumViewModels |> Seq.iter (fun albumVm -> this.Albums.Add albumVm)
    
            let! _ =
                albumViewModels
                |> Seq.map _.LoadCover()
                |> Task.WhenAll
                
            return ()
        }


// NOTE: Instead of a separate AddAlbumIsRunning property (which would require raising PropertyChanged yourself),
// bind directly to AddAlbumCommand.IsRunning in XAML: {Binding AddAlbumCommand.IsRunning}
// Avalonia's binding engine can traverse the nested property path.
